File: 00-每日一题/20200217_75.cpp
//
// Created by zanbo on 2020/2/17.
//

class Solution {
public:
    void sortColors(vector<int>& nums) {
        partition(nums,0,nums.size()-1);
    }

    void partition(vector<int>& nums,int l, int r)
    {
        if(l>=r) return;
        int i=l;
        int key = nums[l];
        for(int j=l+1;j<=r;j++)
        {
            if(nums[j]<=key)
                swap(nums[j],nums[++i]);
        }
        swap(nums[l],nums[i]);
        partition(nums,l,i-1);
        partition(nums,i+1,r);
    }

};

File: 00-每日一题/20200217_75.py
# 三路快排
class Solution:
    def sortColors(self, nums):
        length = len(nums)
        # 三部分， 2条线
        first = -1
        second = length
        i = 0
        while i < second:

            if nums[i] == 0:
                first += 1
                nums[first], nums[i] = nums[i], nums[first]
                i += 1
            elif nums[i] == 1:
                i += 1
            else:
                second -= 1
                nums[second], nums[i] = nums[i], nums[second]
                # 换过来的不知道，所以这边不能懂，要继续判断

        return nums

if __name__ == '__main__':
    obj = Solution()
    while True:
        nums_str = input().strip().split()
        nums = list(map(int, nums_str))
        res = obj.sortColors(nums)
        print(res)